modules/setting/markup.go

```diff
@@ -25,42 +28,114 @@ type MarkupParser struct {
 	IsInputFile    bool
 }
 
+// MarkupSanitizerRule defines the policy for whitelisting attributes on
+// certain elements.
+type MarkupSanitizerRule struct {
+	Element   string
+	AllowAttr string
+	Regexp    *regexp.Regexp
+}
+
 func newMarkup() {
-	extensionReg := regexp.MustCompile(`\.\w`)
 	for _, sec := range Cfg.Section("markup").ChildSections() {
 		name := strings.TrimPrefix(sec.Name(), "markup.")
 		if name == "" {
 			log.Warn("name is empty, markup " + sec.Name() + "ignored")
 			continue
 		}
 
-		extensions := sec.Key("FILE_EXTENSIONS").Strings(",")
-		var exts = make([]string, 0, len(extensions))
-		for _, extension := range extensions {
-			if !extensionReg.MatchString(extension) {
-				log.Warn(sec.Name() + " file extension " + extension + " is invalid. Extension ignored")
-			} else {
-				exts = append(exts, extension)
-			}
+		if name == "sanitizer" {
+			newMarkupSanitizer(name, sec)
+		} else {
+			newMarkupRenderer(name, sec)
 		}
+	}
+}
+
+func newMarkupSanitizer(name string, sec *ini.Section) {
+	haveElement := sec.HasKey("ELEMENT")
+	haveAttr := sec.HasKey("ALLOW_ATTR")
+	haveRegexp := sec.HasKey("REGEXP")
+
+	if !haveElement && !haveAttr && !haveRegexp {
+		log.Warn("Skipping empty section: markup.%s.", name)
+		return
+	}
+
+	if !haveElement || !haveAttr || !haveRegexp {
```

Comment:
Couldn't `haveRegexp` be ignored as your above code?

Reply:
I'm not sure what you mean by as my above code? The first check is for a completely empty section; there's nothing to do so exit with only a warning.

Per above `config-cheat-sheet` description for `REGEXP`: "Must be present but may be empty for unconditional whitelisting of this attribute."

So if it is present but empty, `haveRegexp = ini.Section.HasKey("REGEXP")` will return `true`, but `pattern = ini.Section.Key("REGEXP").ValueWithShadows()[i] == ""` for some `i`. 

This check is thus for if `ELEMENT=` and `HAVEATTR=` have been specified, but `REGEXP` hasn't been. (Well -- strictly it checks for if any haven't been specified having earlier checked that all haven't been specified). 

This is because we _aren't_ parsing the `ini` ourselves. We only have an external library that parses it for us, and gives us parsed results. Particularly, parsed results without any ordering between keys and without knowledge of where a missing key might be. 

If we allowed:

```ini
[markup.sanitizer]
ELEMENT=span
HAVEATTR=class
```

The user might later add:

```ini
[markup.sanitizer]
ELEMENT=span
HAVEATTR=class
ELEMENT=div
HAVEATTR=class
REGEXP=^(error|warning|info)$
```

We wouldn't be able to tell which element/attr pair the regex belongs to, because the ini module doesn't expose neighbor key information. That'd surprise the user (a working config + another working config should equal a working config, since we say that we support adding configs to get more rules).

So I maintain this is desired behavior unless there's something I'm missing?